src/twisted/internet/test/test_inlinecb.py

```diff
@@ -49,6 +359,406 @@ def f(d):
+            returnValue(x)
+
+        def f2(x):
+            expectations.append(("f2 enter", x))
+            x += 2
+            expectations.append(("f2 exit", x))
+            return x
+
+        @inlineCallbacks
+        def f3(x):
+            expectations.append(("f3 enter", x))
+
+            x = yield f1(x)
+            d = f1(x)
+            d.addCallback(f2)
+            x = yield d
+            x += 4
+
+            expectations.append(("f3 exit", x))
+            returnValue(x)
+
+        res = f3(1)
+        self.runCallbacksOnDeferreds(deferredList)
+
+        self.assertEqual(self.successResultOf(res), 9)
+        self.assertEqual(
+            expectations,
+            [
+                ("f3 enter", 1),
+                ("f1 enter", 1),
+                ("f1 exit", 2),
+                ("f1 enter", 2),
+                ("f1 exit", 3),
+                ("f2 enter", 3),
+                ("f2 exit", 5),
+                ("f3 exit", 9),
+            ],
+        )
+
+    def test_raisesExceptionFromDeferredWithWaitingFirstCallback(self):
+        expectations = []
+
+        # list of deferred to invoke with what results
+        deferredList = []
+
+        class MyException(Exception):
+            pass
+
+        @inlineCallbacks
+        def f2(x):
+            expectations.append(("f2 enter", x))
+
+            d = Deferred()
+            deferredList.append((d, MyException()))
+            x = yield d
+
+            expectations.append(("f2 exit", x))
+            returnValue(x)
+
+        @inlineCallbacks
+        def f3(x):
+            expectations.append(("f3 enter", x))
+
+            with self.assertRaises(MyException):
+                x = yield f2(x)
+            x += 4
+
+            expectations.append(("f3 exit", x))
+            returnValue(x)
+
+        res = f3(1)
+        self.runCallbacksOnDeferreds(deferredList)
+        self.assertEqual(self.successResultOf(res), 5)
+        self.assertEqual(
+            expectations,
+            [
+                ("f3 enter", 1),
+                ("f2 enter", 1),
+                ("f3 exit", 5),
+            ],
+        )
+
+    def test_raisesExceptionFromDeferredWithWaitingSecondCallback(self):
+        expectations = []
+
+        # list of deferred to invoke with what results
+        deferredList = []
+
+        class MyException(Exception):
+            pass
+
+        @inlineCallbacks
+        def f1(x):
+            expectations.append(("f1 enter", x))
+
+            d = Deferred()
+            deferredList.append((d, x))
+            x = yield d
+            x += 1
+
+            expectations.append(("f1 exit", x))
+            returnValue(x)
+
+        @inlineCallbacks
+        def f2(x):
+            expectations.append(("f2 enter", x))
+
+            d = Deferred()
+            deferredList.append((d, MyException()))
+            x = yield d
+
+            expectations.append(("f2 exit", x))
+            returnValue(x)
```

Comment:
It looks like something is no ok here.

These lines are never called.

Reply:
It is expected that these lines are not exercised. I added an explanation and no-cover pragma.